rfcs/profiler.md

```diff
@@ -0,0 +1,144 @@
+# Tech Spec: Performance Evaluation & Monitoring for Single Cell
+
+# Goal
+
+Single Cell datasets and queries run across multiple platforms (Mac OSes, Linux, AWS VMs, etc.) and across multiple layers of software stack (python, C++ and R as of now!). Performance and memory monitoring for such applications is essential because:
+
+- Detection of regression as new features are added is very important
+- Performance comparison across different platforms and languages to spot memory and execution time hot spots for optimization is critical
+- Measuring the system scalability as the workload grows in size is very important
+- Profiling a _suite_ of various operations, which can be run as a unit, reproducibly.
+- ~~Having a tool for customers to monitor and debug their workloads~~
+
+The goal of this process is to provide a multi-layer profiler (consisting of a generic top profiler and multiple custom , language dependent, profilers) to help Single Cell detect potential regression caused by bugs or new releases ~~and also help the customers detect performance issues in their queries~~.
+
+# Terminology & Concepts
+
+**SOMA:** The goal of SOMA (“stack of matrices, annotated”) is a flexible, extensible, and open-source API providing access to annotated, 2D matrix data stored in multiple underlying formats and systems.
+
+**FlameGraph:** A flame graph visualizes a distributed request trace and represents each service call that occurred during the request’s execution path with a timed, color-coded, horizontal bar.
+
+# Requirements
+
+**Basic requirements:**
+
+- Our solution should be able to provide timing metrics for a full run.
+- Our solution should be able to provide memory metrics for a full run.
+- Our solution should be able to be incorporated into a release pipeline to protect against / debug regressions in code
+
+**Advanced requirements:**
+
+- Our solution should be able to provide a breakdown of how time was spent in different components / queries
+- Our solution should be able to provide a breakdown of how memory was used in different components / queries
+- Our solution should be able to aggregate metrics from multiple components into a roll-up view for internal use (may be useful for the third party as well)
+- ~~[P2] Our solution should be able to be released to clients to help them identify and improve poor performing queries~~
+
+## Future Work
+
+This work is open ended as there is a chance to add more and more custom profilers to the system. Also right now, we plan to use flamegraph to connect custom profilers and the generic main profiler. This can also be extended by using different intermediate formats or objects.
```

Comment:
```suggestion
This work is open-ended as there is a chance to add more and more custom profilers to the system. Also right now, we plan to use flamegraph to connect custom profilers and the generic main profiler. This can also be extended by using different intermediate formats or objects.
```